### ISSUE TITLE
Report status of screen curtain at startup when "Play a sound when toggling screen curtain" is disabled.

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#10692]
### Steps to reproduce:
1. Open NVDA's vision settings.
2. Check Make screen black.
3. Uncheck "Play sound when toggling screen curtain".
4. Save settings.
5. Restart NVDA.

### Actual behavior:

When NVDA starts, with the "Play sound" setting disabled, screen curtain is enabled, but no indication is given.

When enabling screen curtain while NVDA is already running with this setting disabled, NVDA reports verbally that the feature has been enabled.

### Expected behavior:

NVDA should report verbally that screen curtain has been enabled when starting up, if screen curtain is enabled and the Play Sound setting is disabled.

#### NVDA installed/portable/running from source:
installed
#### NVDA version:
2019.3beta2
#### Windows version:
10 version 1903

**Comment by CyrilleB79** [comment]
Also, if screen curtain activation fails at NVDA startup, the user is not informed whereas he should.

Steps to reproduce:

0. Press Windows+Escape to exit from Windows Magnifier if it was running
1. Open NVDA's vision settings.
2. Check Make screen black.
3. Uncheck "Play sound when toggling screen curtain".
4. Save settings.
5. Quit NVDA
6. Activate Windows Magnifier and color inversion (press Windows+PavNumPlus and then control+shift+I)
7. Start NVDA again

In this case, screen curtain activation will fail (as described in #10545) but the user should be notified that screen curtain activation has failed.

Note: if you try to reproduce many times theses steps, the color inversion command (control+alt+I) should be pressed only the first time, since this setting will be saved automatically.

**Comment by Adriani90** [comment]
cc: @feerrenrut is this a candidate for P2? Maybe a key stroke to report the screen curtain status would be useful here as well (i.e. nvda+0 for enabling or disabling the screen curtain). This would be more consistent to what we use elsewhere. I.e. the status of the option "moving focus to focusable elements in browse mode" is also not announced at the start but you can check the status with NVDA+8.

**Comment by Adriani90** [comment]
in the case of a key stroke we would not need a reporting at the start of NVDA.

**Comment by feerrenrut** [comment]
I don't think this should be as high as P2. A user can use the gesture to ensure that screen curtain is enabled by toggling it. That said I think there is certainly room for improvement in the UX for screen curtain. I think that hearing the earcon or speech for enable / disable should be configurable. I think it is unnecessary to have both. I think screen curtain should be improved so that regardless of how it is enabled / disabled the user is notified. I would be happy for a command to be added to explicitly check the current status. I don't think this should have a default gesture. I think it would be pretty quick to achieve these goals.

**Comment by feerrenrut** [comment]
I would be happy to take a PR for all or any part. Despite requiring a bit of testing, and lots of permutations of state particularly to "ensure the user is notified in a consistent way regardless of how screen curtain is enabled / disabled", I have labelled it as 'good for new dev' since very little knowledge about NVDA itself is necessary to work on this.

**Comment by christopherpross** [comment]
@feerrenrut would it ok, if I work on this issue? because of the limitation of prs. If so, could you confirm the following requirements:

- the user should be informed, if the screencurtain is enabled on startup, when the sound is enabled or disabled. 
- the reporting should be consistend in all situations and all possible settings regarding sound settings and screencurtain settings
- it should exist a gesture, which is not assigned default, which reports the status of the screencurtain.

is that correct?
Or is that to much? Have I somethink forgoten?

all the best,
Christopher

**Comment by feerrenrut** [comment]
Yep, please go ahead. Please don't forget to carefully test all situations with and without "Always show a warning when enabling screencurtin" enabled.
There a few variations, so I'll be looking for a detailed "testing performed" section of the PR.

**Comment by christopherpross** [comment]
Hey,
so a short question,
should the warning dialog also display on startup of nvda, if the screencurtain is permantently enabled and the warning dlg is activated? If so, have you an idea how to do this without breaking the rules of python (don't return out of an __init__) and without changing the visualprividers framerwork?
all the best,

**Comment by feerrenrut** [comment]
While it probably would be safest to warn for every activation when the warning is enabled, it could be a lot of work, and I don't think it currently does this? 

If you were to try to implement that, my suggestion is to investigate doing something like the following:
- move the current `__init__` code to an `_activate` method and an `_activateWithWarning` method 
- then if the warning option is checked, in the `__init__` use `queueHandler.queueFunction(queueHandler.eventQueue, self._activateWithWarning)` or just call `_activate` directly.

**Comment by christopherpross** [comment]
@feerrenrut Yeah, it doesn't do that at this moment. But correct me, if I am wrong, this would be result in, that if the user toggles the screencurtain over the gesture, the dialog will also happen. But also if the user for example press 2 times the gesture to jump over the temporaly setting. in this case, I think, it could result in 2 dialogs. And so on. This looks not like a clean solution. Or I don't understand the code right now? I only want, that my first contribution here is a clean one.

**Comment by feerrenrut** [comment]
> But also if the user for example press 2 times the gesture to jump over the temporaly setting. in this case, I think, it could result in 2 dialogs.

Yes, you'd have to watch out for this problem. I believe the current implementation works around this successfully. One thing that I think will help, is to separate the concept of "enabling" screen curtain from the objects creation. You might also want to be familiar with #11062 which will result in separating the settings modified by gestures and the GUI.

To start with, you could just focus on the core of this issue. A sound / report when screen curtain is enabled / disabled.

**Comment by christopherpross** [comment]
@feerrenrut  ok, I ignore then the warn setting for now and only fix the two issues there are reported at 1 and 2. A pr will come soon. I hope this is ok.

**Comment by JulienCochuyt** [comment]
@christopherpross, any update?

**Comment by christopherpross** [comment]
@JulienCochuyt sorry, I have Forgotten this issue for Long time. I will fix that  now sorry for the inconvenience.

**Comment by JulienCochuyt** [comment]
@christopherpross,
No problem at all, just a gentle reminder.
Thank you for your work.

**Comment by christopherpross** [comment]
Hey @feerrenrut  and @JulienCochuyt  should I cover also the case, if the screencurtain couldn't be disabled/enabled? If so, should I handle that in the screencurtain or in the visionhandler for all providers? I ask because the idention of this issue is to inform the user in all cases whether the screencurtain is turned off or on.

**Comment by XLTechie** [comment]
Would doing it in the vision handler for all providers, cover the case which 
spauned #11847? That issue was caused by the highlighter being turned off 
automatically when it couldn't be started, with no notice to the user.
(The config is changed, so it doesn't try to turn on again, even if that's how 
it was configured originally)

**Comment by christopherpross** [comment]
￼well, that will cover it too, if I understand the issue correctly.

**Comment by feerrenrut** [comment]
Smaller changes are usually easier to get reviewed and merged. If you can see a way to do this iteratively handling each issue, by the third you'll have a good understanding of what is necessary. I'm looking forward to seeing the draft PR. Thanks for your work here @christopherpross

**Comment by seanbudd** [comment]
related #19441

**Comment by makhlwf** [comment]
Welcome!
I tried to add this feature by modifying core.py in the following way:


try:
    import screenCurtain

    log.debug("Initializing Screen Curtain")

    screenCurtain.initialize()
    if screenCurtain.ScreenCurtain.enabled:
        speech.speakMessage(_("Screen curtain has been enabled..."))
    else:
        speech.speakMessage(_("Screen curtain couldn't be enabled. Ensure that Windows Magnifier is not enabled and try enabling screen curtain manually again."))
except Exception as E:
    log.error(E)


However, I have faced an issue. First things first: now, when running NVDA, theand  screen curtain is enabled, and NVDA will say "screen curtain has been enabled" quickly, not letting NVDA speech even start saying loading nvda.. Also, the NVDA "screen curtain has been enabled" notification gets interrupted quickly when NVDA tries to set the navigator object automatically. So, what do you think could be a good way to solve this problem? ...

**Comment by hwf1324** [comment]
First of all, this should probably be handled in `_screenCurtain.py`.

Secondly, it might be related to speech priority. It should be possible to refer to the implementation of the remote feature, using `ui.delayedMessage` or similar.

**Comment by makhlwf** [comment]
> First of all, this should probably be handled in `_screenCurtain.py`.
> 
> Secondly, it might be related to speech priority. It should be possible to refer to the implementation of the remote feature, using `ui.delayedMessage` or similar.

Thanks for this information. I now have another question. When should the notification be read: after NVDA says 'Loading NVDA, please wait,' or after NVDA finishes reading the automatically focused first navigator object? I'm not sure about this, because this is a very quick process and people may miss the notification.

**Comment by XLTechie** [comment]
You could also trigger it to speak using the postNvdaStartup extensionPoint.

It should definitely speak after "Starting NVDA". But probably before automatic reading, as people are likely to silence that speech and never hear this announcement.
